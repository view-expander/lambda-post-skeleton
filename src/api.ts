import axios from 'axios'

const api = axios.create({
  baseURL: process.env.API_PATH,
  timeout: 180000,
})

export default api
