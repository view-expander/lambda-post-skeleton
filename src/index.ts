import { APIGatewayEvent, APIGatewayProxyResult } from 'aws-lambda'
import createSkeletonImage from './create-skeleton-image'
import postSkeletonImage from './post-skeleton-image'

export async function handler(
  event: APIGatewayEvent
): Promise<APIGatewayProxyResult> {
  try {
    if (
      event.pathParameters?.key === undefined ||
      event.pathParameters.key === null
    ) {
      throw new Error('Object key is required')
    }

    const { key } = event.pathParameters
    const res = await createSkeletonImage(key)
      .then((svg) =>
        postSkeletonImage(svg, key.replace(/^(.+)\.jpg$/, '$1.svg'))
      )
      .catch((err) => {
        throw err
      })

    console.log(res)

    return {
      statusCode: res.$response.httpResponse.statusCode,
      body: JSON.stringify({ key: null }),
    }
  } catch (err) {
    return {
      statusCode: err.response?.status || err.statusCode || 500,
      body: JSON.stringify(err.message),
    }
  }
}
