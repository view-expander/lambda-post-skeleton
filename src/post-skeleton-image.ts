import * as AWS from 'aws-sdk'
const s3 = new AWS.S3()

export default async (svg: string, key: string) => {
  try {
    const Bucket = process.env.BUCKET

    if (Bucket === undefined) {
      throw new Error('Bucket name is required')
    }

    return s3
      .putObject({
        Bucket,
        Key: `skeleton/${key}`,
        Body: svg,
        ContentType: 'image/svg+xml',
      })
      .promise()
  } catch (err) {
    throw err
  }
}
