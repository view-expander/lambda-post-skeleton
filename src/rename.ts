export default (key: string) => {
  const regExp = /^(.+?)\.jpg$/

  if (!regExp.test(key)) {
    throw new Error('Invalid key')
  }

  return key.replace(regExp, '$1.svg')
}
