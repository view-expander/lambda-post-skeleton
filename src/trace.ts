import potrace from 'potrace'

export default (buffer: Buffer) =>
  new Promise<string>((resolve, reject) => {
    return potrace.trace(
      buffer,
      {
        color: '#cccccc',
        turdSize: 10,
      },
      (err, svg) => {
        if (err) {
          reject(err)
          return
        }

        resolve(svg)
      }
    )
  })
