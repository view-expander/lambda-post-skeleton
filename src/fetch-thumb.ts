import api from './api'

export default (key: string) =>
  api.get<ArrayBuffer>(`/photo/${key}`, {
    params: {
      size: 'thumb',
    },
    responseType: 'arraybuffer',
  })
