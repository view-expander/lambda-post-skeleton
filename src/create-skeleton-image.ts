import fetchThumb from './fetch-thumb'
import optimize from './optimize'
import trace from './trace'

export default (key: string) =>
  fetchThumb(key)
    .then((thumb) => trace(Buffer.from(thumb.data)))
    .then(optimize)
