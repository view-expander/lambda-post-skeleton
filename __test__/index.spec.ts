import { handler } from '../src'

describe('handler()', () => {
  it('returns dummy response', async () => {
    expect.assertions(1)

    return expect(handler()).resolves.toEqual({
      statusCode: 200,
      body: JSON.stringify({}),
    })
  })
})
