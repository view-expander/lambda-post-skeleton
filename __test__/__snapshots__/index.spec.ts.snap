// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handler() cacths network error 1`] = `
Object {
  "body": "\\"Network Error\\"",
  "statusCode": 500,
}
`;

exports[`handler() catchs response error 1`] = `
Object {
  "body": "\\"Bad Request\\"",
  "statusCode": 400,
}
`;

exports[`handler() returns response 1`] = `
Object {
  "body": "{\\"key\\":null}",
  "statusCode": 200,
}
`;
